packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -116,261 +94,148 @@ library LibProving {
         TaikoData.Config memory config,
         AddressResolver resolver,
         uint256 blockId,
-        bytes[] calldata inputs
-    ) public {
-        // Check and decode inputs
-        if (inputs.length != 3) revert L1_INPUT_SIZE();
-        TaikoData.Evidence memory evidence = abi.decode(
-            inputs[0],
-            (TaikoData.Evidence)
-        );
-        TaikoData.BlockMetadata memory target = abi.decode(
-            inputs[1],
-            (TaikoData.BlockMetadata)
+        bytes calldata evidenceBytes
+    ) internal {
+        TaikoData.InvalidBlockEvidence memory evidence = abi.decode(
```

Comment:
Brecht, I like the idea of converter an invalid proposal to an L2 empty block, but even an empty block needs a new header I believe. With a proposeBlockInvalid method, the txListProof can simply proving the txList is invalid, but if we want to convert an invalid proposal into an empty block, we also need to construct a `BlockHeader`:

```solidity
struct BlockHeader {
    bytes32 parentHash; // the previous block, this can be known.
    bytes32 ommersHash; // this is known
    address beneficiary; // this is known
    bytes32 stateRoot;
    bytes32 transactionsRoot; // this corresponds to an empty tx tree
    bytes32 receiptsRoot;  // this corresponds to an empty receipts tree
    bytes32[8] logsBloom; // this is also default
    uint256 difficulty;
    uint128 height;
    uint64 gasLimit;
    uint64 gasUsed;
    uint64 timestamp;
    bytes extraData;
    bytes32 mixHash;
    uint64 nonce;
    uint256 baseFeePerGas;
}
```

But we do not know the stateRoot on L1. If the prover provides this data, then the txLIstProof will have to prove it as well. If the above are correct, then I think we should just skip invalid proposals instead of converting them into empty blocks.

Reply:
I think that's fine because we won't want to check that these values in the smart contract because then we would still need special logic on-chain to handle these invalid blocks. Ideally they look exactly the same for the smart contract. And on L2 we do know the previous state root and so for "invalid" blocks we simply enforce the state root to remain the same.

Actually I think it may even be simpler to set the transaction list as empty if the circuit detects an error in the data to really minimize the differences in the circuits for both cases.That will make sure that all the normal logic is still being executed like the anchor transaction (and making sure all block data is correct for the new block hash) which I think would be very nice . Basically we don't even have to see this as a special case anymore. We just have to be able to detect that the data is invalid, and if that happens the tx list is set to an empty list and that's it, everything else remains the same.

Reply:
The only concern is if it's necessary to add that logic into the "good" block circuit. As if we assume that the rational proposers do not submit invalid RLP block. Is it better to keep the invalid proof standalone??

Reply:
BTW: for empty list, do we still include anchor?? If so, the state root changes.

Reply:
> I think that's fine because we won't want to check that these values in the smart contract because then we would still need special logic on-chain to handle these invalid blocks. Ideally they look exactly the same for the smart contract. And on L2 we do know the previous state root and so for "invalid" blocks we simply enforce the state root to remain the same.
> 
> Actually I think it may even be simpler to set the transaction list as empty if the circuit detects an error in the data to really minimize the differences in the circuits for both cases.That will make sure that all the normal logic is still being executed like the anchor transaction (and making sure all block data is correct for the new block hash) which I think would be very nice . Basically we don't even have to see this as a special case anymore. We just have to be able to detect that the data is invalid, and if that happens the tx list is set to an empty list and that's it, everything else remains the same.

Please take a look at this PR which merged `proveBlock` and `proveInvalidBlock` into one.

Reply:
> The only concern is if it's necessary to add that logic into the "good" block circuit. As if we assume that the rational proposers do not submit invalid RLP block. Is it better to keep the invalid proof standalone??

That was also my initial idea, but if you do that then there's lots of other changes in the smart contracts/protocol/circuits that are necessary to achieve that "simpler" solution. Instead the proposed approach only needs some very contained  changes in the circuit to support it, while everything else can remain the same and so we don't end up having a lot of code that is basically not used and is not being tested as well as the happy path. 

> BTW: for empty list, do we still include anchor?? If so, the state root changes.

I would do so yes, because I think it's harder to not actually do it using the new approach (ALL blocks always do the anchor transaction, no special case anymore). An "invalid" block will now be an actual block part of the chain, and will modify the chain as if the block was proposed with an empty tx list.

Reply:
Brecht, so there are two options for us:

Option#1: invalid blocks are just empty blocks (probably with an anchor tx). 
Option#2, invalid blocks are still skipped.

I assume for option#1 we just have one circuit (with different tx count), but will that circuit be larger than the circuit that doesn't support invalid txList? If so, it means for both valid and invalid blocks, the proving cost get higher.

If we go with option#2, then we will have two circuits (different sizes), but they are both smaller than the one for option#1.

So I think the big question is:  how worse is the option#1 circuit in terms of cost as well as L2 block gas limit.

Reply:
Created this PR to peak the changes.

https://github.com/taikoxyz/taiko-mono/pull/13291